Pass on the "Pixels" and "Stacking contexts" section of Chapter 11 (#281)
browserengineering/book: book/visual-effects.md

this book, so we need a new graphics library. Let's use [Skia][skia],
the library that Chromium uses. Unlike Tkinter, Skia doesn't handle
inputs or create graphical windows, so we'll pair it with the
[SDL][sdl] GUI library. We'll also add the [Pillow][pillow] library
for handling images.

[skia]: https://skia.org
[sdl]: https://www.libsdl.org/
[pillow]: https://python-pillow.org

[^tkinter-before-gpu]: That's because Tk, the graphics library that
Tkinter uses, dates from the early 90s, before high-performance
graphics cards and GPUs became widespread.

> Start by installing [Pillow][install-pillow] and [Skia's][skia-python]

Reviewer: ```suggestion
Start by installing [Pillow][install-pillow], [Skia][skia-python]
```

> and [SDL's][sdl-python] Python bindings like so:

Reviewer: ```suggestion
and [SDL][sdl-python] like so:
```

Pixels, Color, Raster
=====================

Skia, like the Tkinter canvas we've been using until now, is a
_rasterization_ library: it converts shapes like rectangles and text
into pixels. Before we move on to Skia's advanced features, let's look
at how rasterization works at a deeper level.

You probably already know that computer screens are a 2D array of
pixels. Each pixel contains red, green and blue lights,[^lcd-design]
or _color channels_, that can shine with an intensity between 0 (off)
and 1 (fully on). By mixing red, green, and blue, which is formally

> known as the [sRGB color space][srgb], any color can be

Reviewer: ```suggestion
known as the [sRGB color space][srgb], any color in that space's _gamut_ can be
```

> called the [sRGB color space][srgb], and web pages use those colors

Reviewer: The set of colors is the gamut, not the color space.

Reviewer: A color space is an "organization of colors" - a system of representing them.

via CSS.[^other-spaces]

> [^lcd-design]: Actually, some screens contain [pixels besides red,

Reviewer: Add a sentence at the beginning, porting over from my text:

"sRGB is a color space defined in the 90s based on [CRT]-technology screens of the time. Since that was
when the web came into existence, CSS used this color space."

Author: I put this into the "other-spaces" footnote just below. Do you think it should be moved to this one?

Reviewer: Oh right. I found that later in my code review and forgot to delete this comment. It's ok where it is now.

    Moreover, different screens can use slightly different reds,
    greens, or blues; professional color designers typically have to
    [calibrate their screen][calibrate] to display colors accurately.
    For the rest of us, the software still communicates with the
    display in terms of standard red, green, and blue colors, and the
    display hardware converts to whatever pixels it uses.
    
[lcd-design]: https://geometrian.com/programming/reference/subpixelzoo/index.php
[calibrate]: https://en.wikipedia.org/wiki/Color_calibration
[srgb]: https://en.wikipedia.org/wiki/SRGB
[CRT]: https://en.wikipedia.org/wiki/Cathode-ray_tube
[color-spec]: https://drafts.csswg.org/css-color-4/

[^other-spaces]: The sRGB color space dates back to [CRT
displays][CRT]. New technologies like LCD, LED, and OLED can display
more colors, so CSS now includes [new syntax][color-spec] for

> expressing these new colors.

Reviewer: ```suggestion
expressing these new colors. All color spaces have a limited gamut of expressible colors.
```

> [color-spec]: https://drafts.csswg.org/css-color-5/

Reviewer: ```suggestion
[color-spec]: https://drafts.csswg.org/css-color-4/
```

> New technologies like LCD, LED, and OLED and new pixel chemistries and

Reviewer: ```suggestion
New technologies like LCD, LED, and OLED```

The job of a rasterization library is to determine the red, green, and
blue intensity of each pixel on the screen, based on the
shapes---lines, rectangles, text---that the application wants to
display. This is already a challenge for a single shape, but with
multiple shapes there's an additional question: what color should the
pixel be when two shapes overlap? So far, our browser has only handled
opaque shapes,[^nor-subpixel] and the answer has been simple: take the
color of the top shape. But now we need more nuance.

[^nor-subpixel]: Nor has our browser yet thought about subpixel
    geometry or anti-aliasing, which also rely on color mixing.

Many objects in nature are partially transparent: frosted glass,
clouds, or colored paper, for example. Looking through one, you see

> two colors *blended* together. That's also why computer screens work:

Reviewer: ```suggestion
multiple colors *blended* together. That's also why computer screens work:
```

> has only handled opaque shapes, and the answer has been simple: take

Reviewer: ```suggestion
has only handled opaque shapes^[And without thinking about subpixel geometry or anti-aliasing...], and the answer has been simple: take
```

> [rounded-rects]: https://www.folklore.org/StoryView.py?story=Round_Rects_Are_Everywhere.txt

Reviewer: This is already linked later in the chapter.

the red, green, and blue lights [blend together][mixing] and appear to
our eyes as another color. Designers use this effect[^mostly-models]
in overlays, shadows, and tooltips, so our browser needs to support
color mixing.

[mixing]: https://en.wikipedia.org/wiki/Color_mixing

[^mostly-models]: Mostly. Some more advanced blending modes are
difficult, or perhaps impossible, in real-world physics.

<a name="alpha"></a>

The most important type of color mixing is transparency. Skia, SDL,
and many other color libraries account for transparency with a fourth

> channel *alpha*.[^alpha-history] An alpha of 0 means the pixel is

Reviewer: ```suggestion
*alpha* channel.[^alpha-history] An alpha of 0 means the pixel is
```

fully transparent (meaning, no matter what the colors are, you can't
see them anyway), and an alpha of 1 means a fully opaque like the ones
we've been working with so far. When a pixel with alpha overlaps

> another pixel, the final color of their two colors.

Reviewer: ```suggestion
another pixel, the final color is a mix of their two colors.
```

[^alpha-history]: Check out this [history of alpha][alpha-history],
written by its co-inventor (and co-founder of Pixar), or read this
[derivation of alpha][alpha-deriv] for how it is computed.

[alpha-history]: http://alvyray.com/Memos/CG/Microsoft/7_alpha.pdf
[alpha-deriv]: https://jcgt.org/published/0004/02/03/paper.pdf

::: {.further}
Screens use red, green, and blue color channels to match the three
types of [cone cells][cones] in a human eye. We take it for granted,
but color standards like [CIELAB][cielab] derive from attempts to
[reverse-engineer human vision][opponent-process]. These cone cells
vary between people: some have [more][tetrachromats] or
[fewer][colorblind] (typically an inherited condition carried on the X
chromosome). Moreover, different people have different ratios of cone
types and those cone types use different protein structures that vary

> the exact frequency of green, red, and blue that they respond to. The

Reviewer: ```suggestion
in the exact frequency of green, red, and blue that they respond to. The
```

human perception of color is a topic that combines software, hardware,
chemistry, biology, and psychology.
:::

[cones]: https://en.wikipedia.org/wiki/Cone_cell
[cielab]: https://en.wikipedia.org/wiki/CIELAB_color_space
[opponent-process]: https://en.wikipedia.org/wiki/Opponent_process
[colorblind]: https://en.wikipedia.org/wiki/Color_blindness
[tetrachromats]: https://en.wikipedia.org/wiki/Tetrachromacy#Humans

Stacking contexts
=================

Color mixing means we need to think carefully about the order of
operations. For example, consider black text on an orange background,
placed semi-transparenly over a white background.[^example-1] The text
is gray while the background is yellow-orange. That's due to blending:
the text and the background are both partially transparent and let
through some of the underlying white:

<div style="opacity: 0.5; background: orange; color: black; font-size: 50px; padding: 15px; text-align: center;flex:1;">Test</div>

But importantly, the text isn't orange-gray: even though the text is
partially transparent, none of the orange shines through. That's
because the order matters. First, the text is blended with the

> background; since the text is opaque, its pixels are black, while the

Reviewer: ```suggestion
background; since the text is opaque, its blended pixels are black and overwrite the orange, while the
```

rest of the background is orange. Then, this black-and-orange is
blended with the white background. Doing the operations in a different
order would lead to dark-orange or black text.

> To handle this properly, computers apply blending not to individual

Reviewer: ```suggestion
To handle this properly, browsers apply blending not to individual
```

shapes but to a tree of [*stacking contexts*][stacking-context]. Each
stacking context is rastered into a single 2D array of pixels and then
blended into its parent stacking context. Note that to raster a

> stacking context you first need to raster in and child stacking

Reviewer: ```suggestion
stacking context you first need to raster it *and* its child stacking
```

contexts, so they can be blended into the stacking context. In other
words, rastering a web page requires a bottom-up traversal of the tree
of stacking contexts.

[stacking-context]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context

> In real browsers, stacking contexts are created by HTML elements with

Reviewer: ```suggestion
In real browsers, stacking contexts are formed by HTML elements with
```

certain styles, up to any descendants that themselves have such
styles. The full definition is actually quite complicated, so in this

> chapter we'll simplify by treating every layout object a stacking

Reviewer: ```suggestion
chapter we'll simplify by treating every layout object as a stacking
```

context.

::: {.further}

> Mostly, element [induce a stacking context][stacking-context] because

Reviewer: ```suggestion
Mostly, elements [form a stacking context][stacking-context] because
```

of CSS properties that have something to do with layering (like
`z-index`) or visual effects (like `mix-blend-mode`). On the other
hand, the `overflow` property, which can make an element scrollable,
does not induce a stacking context, which I think was a

> mistake.[^also-containing-block] Inside a modern browser, scrolling is

Reviewer: ```suggestion
mistake.[^also-containing-block] The reason is that inside a modern browser, scrolling is
```

browser, scrolling is done on the GPU by offsetting two surfaces.
Without a stacking context the browser might (depending on the web
page structure) have to move around multiple independent surfaces with

> complex paint orders, in lockstep, to achieve scrolling.

Reviewer: ```suggestion
complex paint orders, in lockstep, to achieve scrolling. Fixed- and sticky-positioned elements also form a stacking
context because of their interaction with scrolling.
```
